backend/inspirehep/pidstore/providers/bai.py

```diff
@@ -80,21 +84,38 @@ def create(
         pid_value = pid_value or get_first_value_for_schema(
             get_value(data, "ids", []), "INSPIRE BAI"
         )
-        if not pid_value:
-            pid_value = cls.generate_bai(data)
-        pid_from_db = cls.query_pid_value(pid_value)
-        if not pid_from_db:
-            pid_from_db = super().create(
-                pid_value=pid_value,
-                object_type=object_type,
-                object_uuid=object_uuid,
-                status=PIDStatus.REGISTERED,
-                **kwargs,
-            )
-        else:
-            if pid_from_db.object_uuid != object_uuid:
-                raise PIDAlreadyExistsError(pid_value=pid_value, pid_type="bai")
-        return pid_from_db
+        retry_count = (
+            current_app.config.get("PIDSTORE_TEXKEY_MAX_RETRY_COUNT", 5)
+            if pid_value
+            else 1
+        )
+        last_exception = None
+        for _ in range(retry_count):
+            new_pid = pid_value or cls.generate_bai(data)
+            pid_from_db = cls.query_pid_value(new_pid)
+            if not pid_from_db:
+                try:
+                    provider_object = super().create(
+                        pid_value=new_pid,
+                        object_type=object_type,
+                        object_uuid=object_uuid,
+                        status=PIDStatus.REGISTERED,
+                        **kwargs,
+                    )
+                    # Pid created successfully.
+                    last_exception = None
+                    break
+                except IntegrityError as e:
+                    last_exception = e
+            elif pid_from_db.object_uuid != object_uuid:
+                last_exception = PIDAlreadyExistsError(
+                    pid_value=pid_value, pid_type="bai"
+                )
+            time.sleep(current_app.config.get("PIDSTORE_TEXKEY_RETRY_DELAY", 5))
+        if last_exception:
+            raise last_exception
+
+        return provider_object
```

Comment:
could we use `backoff` for example, instead of manually handle that?

Reply:
WDYM? Minters do not have separate task.

Reply:
Do you mean installing another package for retrying functions which will be used only in here?

Reply:
don't we have [backoff](https://pypi.org/project/backoff/) already?

Reply:
no https://github.com/inspirehep/inspirehep/blob/master/backend/poetry.lock

Reply:
I was thinking about it but new package to be used only in one place... It's too much I believe.

Reply:
:+1: weird, I thought that we're using it. At least we should add tests for this additional functionality.